### ISSUE TITLE
Bug: useCharacterLimit doesn't work if adding a character before the overflow

### ISSUE THREAD

**Issue by francistogram** [facebook/lexical#5096]
Lexical version:

## Steps To Reproduce
1. Go to https://playground.lexical.dev/?isCharLimit=true the character limit is set to 5
2. Type "abcde" and see the 0 there correctly
<img width="184" alt="image" src="https://github.com/facebook/lexical/assets/3721449/c44c0e25-f81f-46e6-a9cf-9782fb7f8c0d">

3. Type "f"  and see the -1 
<img width="164" alt="image" src="https://github.com/facebook/lexical/assets/3721449/37aa2d8e-3d9f-464d-82b3-2fde8da9ed9f">

4. Add a "1" to the front and see a -4 even when it should be -2
<img width="251" alt="image" src="https://github.com/facebook/lexical/assets/3721449/4902e234-fa6c-4f92-a7d4-8701d2d2e02a">

You also see that the tree structure is 
```
 root
  └ (47) paragraph  
>   ├ (59) text  "1abcd"
    ├ (65) overflow  
    | └ (64) text  "e"
    └ (63) overflow  
      └ (62) text  "f"
```
when the two overflow nodes should be merged and I think the root cause of the bug

Link to code example: https://playground.lexical.dev/?isCharLimit=true

## The current behavior
Does not merge overflow nodes when characters are added before the current overflow and overcounts the characters

## The expected behavior
Should act the same adding characters before or after the overflow

**Comment by JohnyKapo** [comment]
Hello may I work on this issue and get assigned?

**Comment by francistogram** [comment]
If you think you can resolve go for it!

My suspicion is that the fix is somewhere in here 

https://github.com/facebook/lexical/blob/9a679ecb2fa3ca999772e2a170f4ad15d19d4635/packages/lexical-react/src/shared/useCharacterLimit.ts#L143

**Comment by francistogram** [comment]
For posterity I figured out a fix

Take `mergePrevious` https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L244 and instead write `mergeNext` using `const nextNode = overflowNode.getNextSibling();` 

and then replace the call [here](https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L219)

**Comment by JohnyKapo** [comment]
> For posterity I figured out a fix
> 
> Take `mergePrevious` https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L244 and instead write `mergeNext` using `const nextNode = overflowNode.getNextSibling();`
> 
> and then replace the call [here](https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L219)

I've looked into this fix and yes it fixes the number of overflow characters. But it also creates new bug, where again when you add new characters before overflow it shuffles characters (it seems like when you type "12345" and then add 2 new characters '1' at the start of the string it shuffles string to "1112354" where 5 and 4 switched). I will look further into this bug.

**Comment by francistogram** [comment]
Damn it, you're right things get shuffled :(

I don't think I have the permissions to assign an issue in this repo unfortunately otherwise I would

Let me know if you figure out a resolution!

**Comment by francistogram** [comment]
Figured it out

Last thing is also changing `.insertBefore` to `.insertAfter`

https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L259

**Comment by JohnyKapo** [comment]
> Figured it out
> 
> Last thing is also changing `.insertBefore` to `.insertAfter`
> 
> https://github.com/facebook/lexical/blob/main/packages/lexical-react/src/shared/useCharacterLimit.ts#L259

Yes, I've also found about that `insertAfter` function yesterday. It initially fixes the problem, but when you run tests, the only ones that don't go through are those for `useCharacterLimit`. I checked them only briefly and it seems like the first `except` that is failing is due to not merging nodes correctly. I will look deeper into that and give you answer in upcoming hours.

**Comment by francistogram** [comment]
I wonder if the tests were written to test for the incorrect behavior though

Just double check that it actually solves the problem (as it seems like it does) because if it does then it may be that the tests are the ones that are wrong and need to be fixed

**Comment by JohnyKapo** [comment]
> I wonder if the tests were written to test for the incorrect behavior though
> 
> Just double check that it actually solves the problem (as it seems like it does) because if it does then it may be that the tests are the ones that are wrong and need to be fixed

Yep I'm gonna double check if it works in various scenarios and if it does I'll fix tests and make a pull request.

**Comment by francistogram** [comment]
Turns out the solution I suggested causes some issues with pasting and does not merge the overflow correctly in that state

**Comment by JohnyKapo** [comment]
Copying text is kinda weird. I've checkedit and I am pretty sure it has nothing to do with useCharacterLimit.ts (I've gathered that from logs when useCharacterLimit functions were called and it shown that **root** node was already bugged even before merging anything). Everything other works as it should (just that copying is weird). So when I finish this issue I will create PR and probably new issue for this bug.

PS: Also you can try copying something like URL from webpage and it copies/pastes it just fine. :)

**Comment by francistogram** [comment]
It's definitely from `useCharacterLimit` you don't see the issue from the URL because it only happens with new lines it seems

I updated `$wrapOverflowedNodes` to look like this

```typescript
function $wrapOverflowedNodes(offset: number): void {
  let dfsNodes = $dfs();
  const dfsNodesLength = dfsNodes.length;
  let accumulatedLength = 0;

  for (let i = 0; i < dfsNodesLength; i += 1) {
    const { node } = dfsNodes[i];

    if ($isOverflowNode(node)) {
      ...
    }  else if ($isLeafNode(node)) {
      const previousAccumulatedLength = accumulatedLength;
      accumulatedLength += node.getTextContentSize();

      if (accumulatedLength > offset && !$isOverflowNode(node.getParent())) {
        const previousSelection = $getSelection();
        let overflowNode;

        let firstCase = false;

        // For simple text we can improve the limit accuracy by splitting the TextNode
        // on the split point
        if (
          previousAccumulatedLength < offset &&
          $isTextNode(node) &&
          node.isSimpleText()
        ) {
          const [, overflowedText] = node.splitText(
            offset - previousAccumulatedLength
          );
          overflowNode = $wrapNode(overflowedText);
          firstCase = true;
        } else {
          overflowNode = $wrapNode(node);
        }

        if (previousSelection !== null) {
          $setSelection(previousSelection);
        }

        if (!firstCase || accumulatedLength - 1 > offset) {
          mergePrevious(overflowNode);
        } else {
          mergeNext(overflowNode);
        }
      }
    }
```

and then `mergeNext` looks like this and seems to work

```typescript
export function mergeNext(overflowNode: OverflowNode): void {
  const nextNode = overflowNode.getNextSibling();

  if (!$isOverflowNode(nextNode)) {
    return;
  }

  const firstChild = overflowNode.getFirstChild();
  const nextNodeChildren = nextNode.getChildren();
  const nextNodeChildrenLength = nextNodeChildren.length;

  if (firstChild === null) {
    overflowNode.append(...nextNodeChildren);
  } else {
    for (let i = 0; i < nextNodeChildrenLength; i++) {
      firstChild.insertAfter(nextNodeChildren[nextNodeChildrenLength - 1 - i]);
    }
  }

  const selection = $getSelection();

  if ($isRangeSelection(selection)) {
    const anchor = selection.anchor;
    const anchorNode = anchor.getNode();
    const focus = selection.focus;
    const focusNode = anchor.getNode();

    if (anchorNode.is(nextNode)) {
      anchor.set(overflowNode.getKey(), anchor.offset, "element");
    } else if (anchorNode.is(overflowNode)) {
      anchor.set(
        overflowNode.getKey(),
        nextNodeChildrenLength + anchor.offset,
        "element"
      );
    }

    if (focusNode.is(nextNode)) {
      focus.set(overflowNode.getKey(), focus.offset, "element");
    } else if (focusNode.is(overflowNode)) {
      focus.set(
        overflowNode.getKey(),
        nextNodeChildrenLength + focus.offset,
        "element"
      );
    }
  }

  nextNode.remove();
}```

**Comment by JohnyKapo** [comment]
Well I've tried your code and I still had issues with copying. I haven't touched `$wrapOverflowedNodes` and I've managed to fix it by updating function `mergeNext` like this: 

`export function mergeNext(overflowNode: OverflowNode): void {

  let paragraph = overflowNode.getParent();
  let paragraphNextSibling;

  if (paragraph !== undefined) {
    paragraphNextSibling = paragraph?.getNextSibling();
  }

  let nextOverflowNode;
  const formerOverflowNode = overflowNode.getNextSibling();
  if ($isOverflowNode(paragraphNextSibling)) {             // Check if next parent sibling is overflow node (cause of the bug)
    nextOverflowNode = paragraphNextSibling;            
  } else if ($isOverflowNode(formerOverflowNode)) {        // If next parent sibling is not overflow node check if the overflow node is not in current parent node (normal use)
    nextOverflowNode = formerOverflowNode;
  } else {
    return;                                                // There's no overflow node and thus no merging is required
  }

  const nextOverflowNodeChildren = nextOverflowNode.getChildren();
  const nextOverflowNodeChildrenLength = nextOverflowNodeChildren.length;

  for (let i = 0; i < nextOverflowNodeChildrenLength; i++) {
      overflowNode.append(nextOverflowNodeChildren[i]);
  }
  
  const selection = $getSelection();

  if ($isRangeSelection(selection)) {
    const anchor = selection.anchor;
    const anchorNode = anchor.getNode();
    const focus = selection.focus;
    const focusNode = anchor.getNode();

    if (anchorNode.is(nextOverflowNode)) {
      anchor.set(overflowNode.getKey(), anchor.offset, "element");
    } else if (anchorNode.is(overflowNode)) {
      anchor.set(
        overflowNode.getKey(),
        nextOverflowNodeChildrenLength + anchor.offset,
        "element"
      );
    }

    if (focusNode.is(nextOverflowNode)) {
      focus.set(overflowNode.getKey(), focus.offset, "element");
    } else if (focusNode.is(overflowNode)) {
      focus.set(
        overflowNode.getKey(),
        nextOverflowNodeChildrenLength + focus.offset,
        "element"
      );
    }
  }

  nextOverflowNode.remove();
}
`

The most noticeable change is at the start where I've added check of next parent sibling, which in our case of copying text is always `overflow`. If it is `overflow` then function works with next parent sibling instead of next sibling in current parent.
